adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+            original_perception_graph = PerceptionGraph.from_frame(
+                perception.frames[0]
+            ).copy_as_digraph()
+        else:
+            raise RuntimeError("Cannot process perception type.")
+        # Remove learner from the perception
+        observed_perception_graph = graph_without_learner(original_perception_graph)
+        observed_linguistic_description = (
+            learning_example.linguistic_description.as_token_sequence()
+        )
+
+        self.learn_propose_but_verify(
+            observed_perception_graph, observed_linguistic_description
+        )
+
+    # learn method
+    def learn_propose_but_verify(
+        self,
+        observed_perception_graph: PerceptionGraph,
+        observed_linguistic_description: Tuple[str, ...],
+    ) -> None:
+        logging.info(f"Pursuit learner observing {observed_linguistic_description}")
+        for word in observed_linguistic_description:
+            if word in ("a", "the"):
+                continue
+            logging.info(f"Considering '{word}'")
+            if word not in self._words_to_hypotheses:
+                # a) If word has never been seen before, go through initialization step
+                self.initialization_step(word, observed_perception_graph)
+            else:
+                # b) If word has been seen before, test the recorded hypothesis
+                if self.test_hypothesis_step(word, observed_perception_graph):
+                    # If hypothesis is confirmed, do nothing
+                    continue
+                else:
+                    # If hypothesis is not confirmed, treat word as an unseen word, and go
+                    # through the initialization step again (to randomly select meaning)
+                    self.initialization_step(word, observed_perception_graph)
+            if self._log_word_hypotheses_to:
+                self._log_hypotheses(word)
+
+    def initialization_step(self, word: str, observed_perception_graph: PerceptionGraph):
+        # If the word has never been seen before OR word does not match any
+        # 'object' in the current frame, learn a new hypothesis/pattern
+        # generated as a pattern graph from the perception.
+        meanings = [
+            PerceptionGraphPattern.from_graph(object_).perception_graph_pattern
+            for object_ in self.get_objects_from_perception(observed_perception_graph)
+        ]
+
+        #  Pick a random meaning for our hypothesis (Independent of any prior data)
+        pattern_hypothesis = self._rng.choice(meanings)
+
+        if self._graph_logger:
+            self._graph_logger.log_graph(
+                pattern_hypothesis,
+                logging.INFO,
+                "Initializing (randomly chosen) meaning for %s ",
+                word,
+            )
+
+        self._words_to_hypotheses[word] = pattern_hypothesis
+
+    # Test Hypothesis
+    def test_hypothesis_step(
```

Comment:
Rename to "verify_hypothesis()" perhaps? It seems to function as a verification returning a boolean instead of an entire learning step

Reply:
Renamed. I wasn't sure if a naming convention (i.e: with the word 'step')